virttest/libvirt_xml/devices/redirdev.py

```diff
@@ -4,13 +4,60 @@
 http://libvirt.org/formatdomain.html#elementsRedir
 """
 
-import six
+from virttest.libvirt_xml import accessors
+from virttest.libvirt_xml.devices import base, librarian
 
-from virttest.libvirt_xml.devices import base
 
-
-@six.add_metaclass(base.StubDeviceMeta)
 class Redirdev(base.TypedDeviceBase):
-    # TODO: Write this class
-    _device_tag = 'redirdev'
-    _def_type_name = 'spicevmc'
+
+    __slots__ = ('type', 'bus', 'source', 'protocol', 'boot', 'alias', 'address')
+
+    def __init__(self, type_name="redirdev", virsh_instance=base.base.virsh):
+        super(Redirdev, self).__init__(device_tag='redirdev',
+                                       type_name=type_name,
+                                       virsh_instance=virsh_instance)
+        accessors.XMLAttribute('type', self, parent_xpath='/',
+                               tag_name='redirdev', attribute='type')
+        accessors.XMLAttribute('bus', self, parent_xpath='/',
+                               tag_name='redirdev', attribute='bus')
+        accessors.XMLElementNest('source', self, parent_xpath='/',
+                                 tag_name='source', subclass=self.Source,
+                                 subclass_dargs={
+                                     'virsh_instance': virsh_instance})
+        accessors.XMLElementDict('protocol', self, parent_xpath='/',
+                                 tag_name='protocol')
+        accessors.XMLElementDict('boot', self, parent_xpath='/',
+                                 tag_name='boot')
+        accessors.XMLElementDict('alias', self, parent_xpath='/',
+                                 tag_name='alias')
+        accessors.XMLElementNest('address', self, parent_xpath='/',
+                                 tag_name='address', subclass=self.Address,
+                                 subclass_dargs={'type_name': 'usb',
+                                                 'virsh_instance': virsh_instance})
+    Address = librarian.get('address')
+
+    def new_source(self, **dargs):
+        """
+        Return a new Redirdev Source instance and set properties from dargs
+        """
+        new_one = self.Source(virsh_instance=self.virsh)
+        for k, v in dargs.items():
+            if v:
+                setattr(new_one, k, v)
+        return new_one
+
+    class Source(base.base.LibvirtXMLBase):
```

Comment:
Since redir dev could be used as **unix** type:
```xml
<redirdev bus="usb" type="unix">
      <source mode="connect" path="/tmp/connect.socket" />
      <protocol type="raw" />
    </redirdev>
```
It is more perfect to add **source->code** in avocado-vt although it is seldomly used in  libvirt.

Reply:
updated